#**Numpy and Pandas Project:**

##Salaries of Different Data Science Fields in the Data Science Domain






###Data Science Job Salaries Dataset contains 11 columns, each are:

1. work_year: The year the salary was paid.
2. experience_level: The experience level in the job during the year
3. employment_type: The type of employment for the role
4. job_title: The role worked in during the year.
5. salary: The total gross salary amount paid.
6. salary_currency: The currency of the salary paid as an ISO 4217 currency code.
7. salaryinusd: The salary in USD
8. employee_residence: Employee's primary country of residence in during the work year as an ISO 3166 country code.
9. remote_ratio: The overall amount of work done remotely
10. company_location: The country of the employer's main office or contracting branch
11. company_size: The median number of people that worked for the company during the year


This dataset was taken from kaggle: https://www.kaggle.com/datasets/arnabchaki/data-science-salaries-2023?resource=download

In [2]:
# Importing the libraries

import numpy as np
import pandas as pd

In [3]:
# Reading the dataset from the given location in the drive

df = pd.read_csv('/content/drive/MyDrive/Platzi_Courses/Pandas_Numpý/ds_salaries.csv')

In [4]:
df

work_year experience_level employment_type                 job_title  \
0          2023               SE              FT  Principal Data Scientist   
1          2023               MI              CT               ML Engineer   
2          2023               MI              CT               ML Engineer   
3          2023               SE              FT            Data Scientist   
4          2023               SE              FT            Data Scientist   
...         ...              ...             ...                       ...   
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   
3754       2021               SE              FT      Data Science Manager   

       salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0       80000             EUR          85847                 ES           100   
1       30000             USD          30000                 US           100   
2       25500             USD          25500                 US           100   
3      175000             USD         175000                 CA           100   
4      120000             USD         120000                 CA           100   
...       ...             ...            ...                ...           ...   
3750   412000             USD         412000                 US           100   
3751   151000             USD         151000                 US           100   
3752   105000             USD         105000                 US           100   
3753   100000             USD         100000                 US           100   
3754  7000000             INR          94665                 IN            50   

     company_location company_size  
0                  ES            L  
1                  US            S  
2                  US            S  
3                  CA            M  
4                  CA            M  
...               ...          ...  
3750               US            L  
3751               US            L  
3752               US            S  
3753               US            L  
3754               IN            L  

[3755 rows x 11 columns]

In [5]:
df.head(5)

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

We have already created a dataframe from the given the dataset. Now we can start applying some Pandas' functions to learn more about our data.



First, lets get an overview of our dataframe, specifically, lets find out the basic information about each row with the `df.info()` function.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


As we can see, we have 3755 rows and 11 columns. The columns `work_year, salary, salary_in_usd` and `remote_ratio` are `int` (numeric) type and the remaining are `object` (text).



Lets investigate about null values in our dataset. Thus, we will use the `df.isnull()` function.

In [7]:
df.isnull()

work_year  experience_level  employment_type  job_title  salary  \
0         False             False            False      False   False   
1         False             False            False      False   False   
2         False             False            False      False   False   
3         False             False            False      False   False   
4         False             False            False      False   False   
...         ...               ...              ...        ...     ...   
3750      False             False            False      False   False   
3751      False             False            False      False   False   
3752      False             False            False      False   False   
3753      False             False            False      False   False   
3754      False             False            False      False   False   

      salary_currency  salary_in_usd  employee_residence  remote_ratio  \
0               False          False               False         False   
1               False          False               False         False   
2               False          False               False         False   
3               False          False               False         False   
4               False          False               False         False   
...               ...            ...                 ...           ...   
3750            False          False               False         False   
3751            False          False               False         False   
3752            False          False               False         False   
3753            False          False               False         False   
3754            False          False               False         False   

      company_location  company_size  
0                False         False  
1                False         False  
2                False         False  
3                False         False  
4                False         False  
...                ...           ...  
3750             False         False  
3751             False         False  
3752             False         False  
3753             False         False  
3754             False         False  

[3755 rows x 11 columns]

This way of exploring the null values does not give us much information, since it's hard to see if there are actually any null values in any colmun. Thus, we can use the aggregate function `df.sum()` to get the total amount of null entries per column.

In [8]:
df.isnull().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

Furtunetely, our dataset does not contain any null entries, as shown in the result above.

Now let's explore about the different and unique values of `experience_level,	employment_type,	job_title`.


 To do so, we select each column as a pandas series and apply the `unique()` function. This returns the unique values in each column as an array. Even more, the function `nunique()` returns the number of unique values.

In [9]:
df['experience_level'].unique()

array(['SE', 'MI', 'EN', 'EX'], dtype=object)

In [10]:
df['employment_type'].unique()

array(['FT', 'CT', 'FL', 'PT'], dtype=object)

In [11]:
df['job_title'].unique()

array(['Principal Data Scientist', 'ML Engineer', 'Data Scientist',
       'Applied Scientist', 'Data Analyst', 'Data Modeler',
       'Research Engineer', 'Analytics Engineer',
       'Business Intelligence Engineer', 'Machine Learning Engineer',
       'Data Strategist', 'Data Engineer', 'Computer Vision Engineer',
       'Data Quality Analyst', 'Compliance Data Analyst',
       'Data Architect', 'Applied Machine Learning Engineer',
       'AI Developer', 'Research Scientist', 'Data Analytics Manager',
       'Business Data Analyst', 'Applied Data Scientist',
       'Staff Data Analyst', 'ETL Engineer', 'Data DevOps Engineer',
       'Head of Data', 'Data Science Manager', 'Data Manager',
       'Machine Learning Researcher', 'Big Data Engineer',
       'Data Specialist', 'Lead Data Analyst', 'BI Data Engineer',
       'Director of Data Science', 'Machine Learning Scientist',
       'MLOps Engineer', 'AI Scientist', 'Autonomous Vehicle Technician',
       'Applied Machine Learning Sc

In [12]:
df['job_title'].nunique()

93

As we can see, there are four different level of experience level:

1. EN = Entry-level
2. MI = Mid-level
3. SE = Senior
4. EX = Executive

Also four different types of employment:

1. FT = Full Time
2. CT = Contractor
3. FL = Freelancer / Independent
4. PT = Part Time

Also, our dataset contains information about 93 different roles in the Data Science domain.

The columns  `salary` and `salary_currency` are not very important for the aim of this project, we mainly focus on the salary in USD. Thus let's delete such columns.

In [13]:
df = df.drop(['salary','salary_currency'],axis=1)
df

work_year experience_level employment_type                 job_title  \
0          2023               SE              FT  Principal Data Scientist   
1          2023               MI              CT               ML Engineer   
2          2023               MI              CT               ML Engineer   
3          2023               SE              FT            Data Scientist   
4          2023               SE              FT            Data Scientist   
...         ...              ...             ...                       ...   
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   
3754       2021               SE              FT      Data Science Manager   

      salary_in_usd employee_residence  remote_ratio company_location  \
0             85847                 ES           100               ES   
1             30000                 US           100               US   
2             25500                 US           100               US   
3            175000                 CA           100               CA   
4            120000                 CA           100               CA   
...             ...                ...           ...              ...   
3750         412000                 US           100               US   
3751         151000                 US           100               US   
3752         105000                 US           100               US   
3753         100000                 US           100               US   
3754          94665                 IN            50               IN   

     company_size  
0               L  
1               S  
2               S  
3               M  
4               M  
...           ...  
3750            L  
3751            L  
3752            S  
3753            L  
3754            L  

[3755 rows x 9 columns]

It would be better if in the columns `experience_level` and `employment_type` we can see the actual names of the categories. Let's do that.

To do so, let's create a dictionary and then through a lambda function, we will replace the names of the categories.

In [14]:
exp_level = {
  'EN' : 'Entry-level',
  'MI' : 'Mid-level',
  'SE' : 'Senior',
  'EX' : 'Executive'
}

emp_type = {
  'FT' : 'Full Time',
  'PT' : 'Part Time',
  'CT' : 'Contractior',
  'FL' : 'Freelancer'
}

In [15]:
df['experience_level'].map(lambda x: exp_level[x])

0            Senior
1         Mid-level
2         Mid-level
3            Senior
4            Senior
           ...     
3750         Senior
3751      Mid-level
3752    Entry-level
3753    Entry-level
3754         Senior
Name: experience_level, Length: 3755, dtype: object

In [16]:
df['employment_type'].map(lambda x: emp_type[x])

0         Full Time
1       Contractior
2       Contractior
3         Full Time
4         Full Time
           ...     
3750      Full Time
3751      Full Time
3752      Full Time
3753    Contractior
3754      Full Time
Name: employment_type, Length: 3755, dtype: object

In [17]:
df['experience_level'] = df['experience_level'].map(lambda x: exp_level[x])
df['employment_type'] = df['employment_type'].map(lambda x: emp_type[x])

In [18]:
df.head()

work_year experience_level employment_type                 job_title  \
0       2023           Senior       Full Time  Principal Data Scientist   
1       2023        Mid-level     Contractior               ML Engineer   
2       2023        Mid-level     Contractior               ML Engineer   
3       2023           Senior       Full Time            Data Scientist   
4       2023           Senior       Full Time            Data Scientist   

   salary_in_usd employee_residence  remote_ratio company_location  \
0          85847                 ES           100               ES   
1          30000                 US           100               US   
2          25500                 US           100               US   
3         175000                 CA           100               CA   
4         120000                 CA           100               CA   

  company_size  
0            L  
1            S  
2            S  
3            M  
4            M

# Problems

Now let's ask couple questions and answer them by the proper use of Pandas' functions.

1. List the top 10 countries with more companies and group them by size.

2. What is the ratio of employees living in a different location from their companies.

3. what is the percentage of employees working remotely, hybrid and on-site.

4. What is the top 5 five roles with more employees in the data science domain and its average salary according to our dataset.

5. From the top five roles with more employes, show the average salary by experience and by employment type.

# First Question

In [19]:
top_10_countries = df[['company_location','job_title']].groupby('company_location').count().sort_values('job_title',ascending=False).head(10)

In [20]:
top_10_countries.columns = ['total_companies']
top_10_countries

total_companies
company_location                 
US                           3040
GB                            172
CA                             87
ES                             77
IN                             58
DE                             56
FR                             34
BR                             15
PT                             14
AU                             14

Thus, we found out that the country with most companies in our dataset is the US and the 10th is Australia. Now let's see how these number of companies is distributed along its sizes.

In [21]:
top_by_size = df[['company_location','company_size','job_title']].groupby(['company_location','company_size']).count().sort_values('job_title',ascending=False).head(10)

In [22]:
top_by_size.columns = ['total_companies_by_size']
top_by_size

total_companies_by_size
company_location company_size                         
US               M                                2723
                 L                                 263
GB               M                                 144
ES               M                                  69
CA               M                                  65
US               S                                  54
IN               L                                  38
DE               M                                  21
GB               L                                  20
FR               M                                  19

This way we see that medium companies from the US are the ones employing the majority of people in our dataset.

# Second Question

The ratio of employees living in a different location from their companies is basically how many employees work for a company based abroad their home country in comparison to those who live and work for companies based in their own coutries. To find this ratio, let's just find the amount of employees for each situation explained above.

First, let's see how many employees live in a different country than the company they work for:

In [23]:
df[['employee_residence','company_location']][df['employee_residence']!=df['company_location']]

employee_residence company_location
81                   PT               US
82                   IN               DE
183                  NG               CA
218                  AU               FI
249                  ES               GB
...                 ...              ...
3729                 IN               AS
3731                 FR               LU
3735                 BR               US
3743                 ES               US
3746                 SG               IL

[96 rows x 2 columns]

by finding the shape of this df, we can find how many entries (employee) satisfy the condition `employee_residence != company_location`.

In [24]:
df[['employee_residence','company_location']][df['employee_residence']!=df['company_location']].shape[0]

96

Thus, 96 employees work for a company based in a different from their home country.

In the same way, we find out the other situation.

In [25]:
df[['employee_residence','company_location']][df['employee_residence']==df['company_location']].shape[0]

3659

Thus, The ratio of employees living in a different location from their companies is:

In [26]:
96/3659

0.02623667668761957

So we can say that around 2.6% of the employees in our dataset work for a company based in a different country than their home one.

# Third Question

To answer this question, let's just simply find how many employees have `remote_ratio` of 100, 50 and 0. Then we just find this quantity by the total of employees in our datarset.

In [27]:
df[['remote_ratio']].value_counts()

remote_ratio
0               1923
100             1643
50               189
dtype: int64

In [28]:
work_model = pd.DataFrame(df[['remote_ratio']].value_counts())

In [29]:
work_model.columns = ['number_employees']
work_model = work_model.rename(index={0:'on_site',100:'remote',50:'hybrid'})
work_model = work_model.rename_axis('work_model')
work_model['percentage'] = (work_model['number_employees']/work_model['number_employees'].sum())*100

In [30]:
work_model

number_employees  percentage
work_model                              
on_site                 1923   51.211718
remote                  1643   43.754993
hybrid                   189    5.033289

Then we can see that more than half of the people work from the office. However, it is important to notice that around 48% work fully remote, which is great news for those (like me) looking for these kind of opportunities.

# Fourth Question

To find the top five data science roles and its average salary, we can just groupby the column `job_title` and then using the aggregation function, we count how many employees are per role and the average salary for each role.

In [58]:
top_5 = df.groupby('job_title').agg({'work_year' : 'count', 'salary_in_usd' : 'mean'}).sort_values('work_year',ascending=False)
top_5.columns = ['number_of_employees','avg_salary']
top_5.head()

number_of_employees     avg_salary
job_title                                                    
Data Engineer                             1040  142785.734615
Data Scientist                             840  140869.765476
Data Analyst                               612  108716.227124
Machine Learning Engineer                  289  154690.726644
Analytics Engineer                         103  152368.631068

As we see, Data Engineer is the role with more employees with a total of 1040 and its average salary is USD \$142,786 per year. Within the top 5 roles with more employees, the Machine Learning Engineers are the employees best paid, with an average salary of USD \$154,691 per year.

# Fifth Question

5. From the top five roles with more employes, show the average salary by experience and by employment type.

To answer this question, let's start from the previous result and just add another groupying by column,`experience_level` and `employment_type`. This way we can see how many employees are per role per experience level and its average salary.

In [81]:
top_5_exp=df.groupby(['job_title','experience_level']).agg({'work_year' : 'count', 'salary_in_usd' : 'mean'}).sort_values('work_year',ascending=False)
top_5_exp = top_5_exp.join(top_5['number_of_employees']) # this is just to be able to sort by the total of employees, we drop it later.
top_5_exp.columns = ['number_of_employees','avg_salary','number_of_employees_total']
top_5_exp=top_5_exp.sort_values('number_of_employees_total',ascending=False).drop('number_of_employees_total',axis=1)
top_5_exp.head(19)

number_of_employees     avg_salary
job_title                 experience_level                                    
Data Engineer             Senior                            718  153210.413649
                          Entry-level                        66   91526.742424
                          Executive                          51  209920.960784
                          Mid-level                         205  106074.892683
Data Scientist            Executive                          12  180645.833333
                          Entry-level                        59   74085.966102
                          Mid-level                         161   93945.819876
                          Senior                            608  158990.960526
Data Analyst              Entry-level                        63   59802.746032
                          Executive                           2  120000.000000
                          Mid-level                         167  101516.622754
                          Senior                            380  119930.215789
Machine Learning Engineer Entry-level                        21   94275.571429
                          Mid-level                          59  121011.847458
                          Senior                            209  170268.583732
Analytics Engineer        Entry-level                         1  130000.000000
                          Mid-level                          13  102480.230769
                          Executive                           8  175125.000000
                          Senior                             81  158404.024691

Here is the same idea, we just group by `employment_type`.

In [83]:
top_5_exp=df.groupby(['job_title','employment_type']).agg({'work_year' : 'count', 'salary_in_usd' : 'mean'}).sort_values('work_year',ascending=False)
top_5_exp = top_5_exp.join(top_5['number_of_employees']) # this is just to be able to sort by the total of employees, we drop it later.
top_5_exp.columns = ['number_of_employees','avg_salary','number_of_employees_total']
top_5_exp=top_5_exp.sort_values('number_of_employees_total',ascending=False).drop('number_of_employees_total',axis=1)
top_5_exp.head(13)

number_of_employees     avg_salary
job_title                 employment_type                                    
Data Engineer             Full Time                       1037  143061.610415
                          Freelancer                         1   20000.000000
                          Part Time                          2   61137.000000
Data Scientist            Full Time                        836  141147.050239
                          Freelancer                         1  100000.000000
                          Part Time                          3   77223.000000
Data Analyst              Contractior                        1   90000.000000
                          Full Time                        606  109292.504950
                          Part Time                          5   42614.600000
Machine Learning Engineer Freelancer                         1  100000.000000
                          Full Time                        288  154880.625000
Analytics Engineer        Full Time                        102  153788.911765
                          Contractior                        1    7500.000000

From the results above we can see two things:

1. Executives make the most money in the data science domain, as expected.

2. The majority of the people work full time for its companies.

# Conclusion


This project was really useful to practice all the pandas knoledge acquired during the Platzi Course "Manipulación y Transformación de Datos con Pandas y Numpy".